### PR TITLE
feat: requireHostDir() helper + arc/host on TestEnv (#117 phase 3a)

### PR DESCRIPTION
**Pull request by jcfischer**
## Summary

Phase 3a of #117 — lands Holly's `requireHostDir()` helper from cycle 1 of #119 and prepares the test surface for the wide PaiPaths rename. **Pure additive, zero behavior change**, every existing test passes unmodified.

I re-scoped Phase 3 mid-flight: the full mechanical rename across 17 files + 30 test files was too large for one safe PR. Splitting into 3a (this PR — helper + test surface) and 3b (wide rename + delete PaiPaths) makes both reviewable.

## Changes

- **`src/lib/hosts/dispatch.ts`** — new `requireHostDir(host, type, description?)` throwing variant of `hostPathFor`. Use at install dispatch points where the type must live in a host directory; the optional `description` argument keeps the "expose a bin directory for skill CLIs" wording at the one site that needs it.
- **`src/lib/artifact-installer.ts`** — collapse 5× repeated `if (!dir) throw` blocks (tool / agent / prompt / skill / skill-CLI) into single `requireHostDir()` calls. Net -15 lines.
- **`test/helpers/test-env.ts`** — `TestEnv` exposes `arc: ArcPaths` and `host: HostAdapter` alongside the legacy `paths`. The combined `paths` is composed from arc + host so existing `env.paths.X` accesses keep working without change. The new fields let phase 3b migrate test sites incrementally rather than flag-day.
- **`test/unit/hosts-dispatch.test.ts`** — 4 new tests pinning `requireHostDir` (passthrough for supported types, default error message, custom description, host id baked in).

## What stays for phase 3b

- Wide rename of remaining ~400 `paths.X` accesses across commands + tests to `arc.X` / `host.paths.X`
- Migrate function signatures to take `arc: ArcPaths, host: HostAdapter` instead of `paths: PaiPaths`
- Delete `PaiPaths` type and `createPaths` shim

3b is mechanical search-and-replace — easier to land and review now that the helper and TestEnv shape are in place.

## Test plan

- [x] `bun test test/unit/hosts-dispatch.test.ts` — 17 pass (13 prior + 4 new)
- [x] `bun test` full suite — **704/704 pass** (700 prior + 4 new)
- [x] Zero call site changes outside `artifact-installer.ts`
- [x] `env.paths.X` still resolves identically to before this PR

## References

- Issue: #117
- Holly's `requireHostDir` suggestion: #119 cycle 1 inline comment on `artifact-installer.ts:260`
- Phase 1: #118 (merged e2df58f)
- Phase 2: #119 (merged 393ffa2)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: test/unit/hosts-dispatch.test.ts [hunk 1]
```diff
@@ -157,3 +157,46 @@ describe("createArtifactSymlinks null-guard throws", () => {
...
+      paths: {
+        root: "",
+        skillsDir: "",
+        agentsDir: "",
+        promptsDir: "",
+        binDir: "",
+        settingsPath: "",
+      },
+      supports: () => false,
+    };
```
Comment (jcfischer) [inline test/unit/hosts-dispatch.test.ts]: **[nit/duplication]** This `stubHost` is identical to `makeEmptyPathHost()` defined at line 60 in this same file. Should reuse the factory instead of hand-rolling an identical object.

```typescript
const stubHost = makeEmptyPathHost();
expect(() => requireHostDir(stubHost, "skill")).toThrow(/^Host claude-code/);
```

File: src/lib/hosts/dispatch.ts [hunk 2]
```diff
@@ -57,3 +57,28 @@ export function hostPathFor(
...
+ * Throwing variant of {@link hostPathFor}. Use at install/remove dispatch
+ * points that have already established the artifact type *must* live in a
+ * host directory (skill / agent / prompt / tool) — turning a null into a
+ * single-line precondition error keeps the call sites readable.
+ *
+ * For types that legitimately return null (component / rules / library /
+ * pipeline / action), use {@link hostPathFor} directly and branch on null.
+ *
+ * Suggested by Holly in cycle 1 of #119; landed in Phase 3 of #117 alongside
+ * the wide rename.
```
Comment (jcfischer) [inline src/lib/hosts/dispatch.ts]: **[nit/quality]** Provenance tracking ("Suggested by Holly…") belongs in the commit message, not the JSDoc. The functional docs above this line are good — the history reference will rot and confuse future readers who don't know what cycle 1 of #119 was.

Also "alongside the wide rename" is slightly inaccurate — the wide rename is Phase 3b, not this PR.

**Comment by jcfischer** [comment]
## Full Review: the-metafactory/arc#120

### Lenses Applied (6 + duplication)
| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 0 | 0 | 1 | pass |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass (n/a — no API surface) |
| Architecture | 0 | 0 | 0 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass (n/a — pure function wrapper) |
| Duplication | 0 | 0 | 0 | 1 | pass |

### What I checked

**Correctness.** Traced every `requireHostDir` call site in `artifact-installer.ts` against the old inline `if (!dir) throw` guards. All 5 replacements produce identical error messages — except the `skill|system` case, which now correctly says "support system artifacts" for type=system instead of the old hardcoded "skill artifacts". That's an improvement, not a regression. The `description` parameter for the skill-CLI case (`"expose a bin directory for skill CLIs"`) produces the same message as before.

**TestEnv composition.** The new `arc + host → paths` composition in `test-env.ts` is sound. `createPaths({...})` internally calls `createArcPaths(overrides)`, so `pipelinesDir` (not explicitly passed) defaults correctly from `configRoot`. `ensureDirectories(paths)` still receives a full `PaiPaths` with both arc and host directories. The `openDatabase(arc.dbPath)` switch from `openDatabase(paths.dbPath)` resolves to the same path since `paths.dbPath` is derived from `arc.dbPath`.

**Back-compat.** `env.paths.X` accesses still work because the legacy `PaiPaths` is composed from `arc` + `host` fields. The `@deprecated` tag on `TestEnv.paths` correctly references #117 and signals removal timeline. Phase 3b can delete it without ambiguity.

**Guard edge cases.** `requireHostDir` checks `if (!dir)` which catches both `null` (from `hostPathFor` for non-host types) and `""` (from stub adapters with empty paths). Existing test at line 76 already verified this falsy-guard behavior. The 4 new tests cover: happy path passthrough, default error message, custom description, and host ID embedding.

**Hidden coupling.** No new cross-module dependencies. `artifact-installer.ts` already imported from `hosts/dispatch.ts`; the import just changed from `hostPathFor` to `requireHostDir`.

**Reversibility.** Pure additive + refactor. No migrations, no schema changes, no deletes. Fully reversible by reverting the commit.

### Findings (inline)

1. **[nit/duplication]** `test/unit/hosts-dispatch.test.ts:186-199` — `stubHost` duplicates `makeEmptyPathHost()` from line 60.
2. **[nit/quality]** `src/lib/hosts/dispatch.ts:71` — provenance note in JSDoc belongs in commit message; "alongside the wide rename" is inaccurate for this PR.

### Verdict
Clean PR. The `requireHostDir` helper eliminates repetitive null-guard boilerplate exactly as suggested. TestEnv expansion is well-structured for incremental migration. Both nits are cosmetic and won't cause bugs.

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — clean helper extraction + well-structured TestEnv expansion. Two cosmetic nits, zero functional issues.

verdict: blockers=0, majors=0, nits=2 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)

File: src/lib/hosts/dispatch.ts [hunk 3]
```diff
@@ -57,3 +57,28 @@ export function hostPathFor(
...
+/**
+ * Throwing variant of {@link hostPathFor}. Use at install/remove dispatch
+ * points that have already established the artifact type *must* live in a
+ * host directory (skill / agent / prompt / tool) — turning a null into a
+ * single-line precondition error keeps the call sites readable.
+ *
+ * For types that legitimately return null (component / rules / library /
+ * pipeline / action), use {@link hostPathFor} directly and branch on null.
+ *
+ * Suggested by Holly in cycle 1 of #119; landed in Phase 3 of #117 alongside
```
Comment (jcfischer) [inline src/lib/hosts/dispatch.ts]: **[nit/quality]** JSDoc says "landed in Phase 3 of #117 alongside the wide rename" but the wide rename is Phase 3b, not this PR (3a). Minor temporal inaccuracy — suggests the rename shipped with this function when it has not yet.

Fix: `"landed in Phase 3a of #117; wide rename follows in 3b."` or just drop the forward reference entirely.

File: src/lib/hosts/dispatch.ts [hunk 4]
```diff
@@ -57,3 +57,28 @@ export function hostPathFor(
...
+ * the wide rename.
+ */
+export function requireHostDir(
+  host: HostAdapter,
+  type: ArtifactType | "system",
+  description?: string,
```
Comment (jcfischer) [inline src/lib/hosts/dispatch.ts:73]: **[nit/quality]** The `description` parameter replaces the predicate in `"Host X does not {description}"`, which means callers must provide a grammatically correct verb phrase that reads after "does not". Implicit contract — a caller writing `requireHostDir(host, "tool", "tool artifacts")` gets "Host claude-code does not tool artifacts".

Only one call site uses it today and it reads perfectly (`"expose a bin directory for skill CLIs"`), so this is fine for now. A one-line `@param` tag would prevent future misuse:
```
@param description  Verb phrase completing "Host {id} does not {description}" (default: "support {type} artifacts")
```

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — clean additive PR, requireHostDir helper is well-designed, TestEnv back-compat shape is correct for incremental migration.

verdict: blockers=0 majors=0 nits=2 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)

**Review by jcfischer — COMMENTED** [review]
Re-review (cycle 2) — both cycle-1 nits resolved cleanly:

1. **Provenance line dropped** — `dispatch.ts` JSDoc no longer references Holly or #119 history. Functional docs remain, rot-prone references gone.
2. **`makeEmptyPathHost` lifted + reused** — factory is now module-scope in `hosts-dispatch.test.ts`, shared between `createArtifactSymlinks null-guard throws` and `requireHostDir` describe blocks. The inline 13-line duplicate is gone; one `makeEmptyPathHost()` call replaces it.

Commit `e7c73b0` is -3 lines in dispatch.ts (pure deletion), -13 net in the test file (duplication removed). No functional change, no new code paths, suite 704/704.

I re-scanned the full diff one more time. The `@param description` tag I mentioned in cycle 1 was explicitly called "fine for now" — still fine, single call site, not worth blocking on.

verdict: blockers=0 majors=0 nits=0 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)

**Review by jcfischer — COMMENTED** [review]
Full review cycle 2 (6 lenses + duplication) — both cycle-1 nits resolved, clean helper extraction, well-structured TestEnv expansion. One cosmetic nit (non-standard commit type prefix).

verdict: blockers=0 majors=0 nits=1 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)
